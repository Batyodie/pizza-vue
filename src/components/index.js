import Button from "./Button/Button.vue";
import Card from "./Card/Card.vue";
import DropDown from "./DropDown/DropDown.vue";
import Logo from "./Logo/Logo.vue";
import ShopCart from "./ShopCart/ShopCart.vue";

export { Button, Card, DropDown, Logo, ShopCart };
