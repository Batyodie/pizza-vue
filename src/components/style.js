import BtnTag from "@/components/Button/Button.css";
import Button from "@/components/Button/Button.css";
import DropDownStyle from "@/components/DropDown/DropDown.css";
import CardStyle from "@/components/Card/Card.css";
import LogoStyle from "@/components/Logo/Logo.css";
import ShopCartStyle from "@/components/ShopCart/ShopCart.css";
import BtnStyle from "@/components/Button/Button.css";

export {
  Button,
  BtnTag,
  DropDownStyle,
  CardStyle,
  LogoStyle,
  ShopCartStyle,
  BtnStyle
};
