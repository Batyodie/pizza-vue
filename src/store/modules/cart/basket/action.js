import getCartData from "@/func/getCartData";
import produce from "immer";
const getCartMethod = getCartData();

export default {
  actions: {
    removePizzaItems({ commit, rootState }) {
      try {
        commit("CLEAR_PIZZA_ITEMS", rootState.cart);
      } catch (err) {
        commit("CLEAR_PIZZA_ITEMS_ERROR", err);
      }
    },
    removePizzaItem({ commit, rootState }, deleteItemID) {
      try {
        const updateCart = produce(rootState.cart, draft => {
          const cartItem = draft.pizzaItems[deleteItemID];

          delete draft.pizzaItems[deleteItemID];
          draft.totalPrice = getCartData().getTotalPrice(
            cartItem.items,
            "price"
          );
          draft.pizzaItemsCount = cartItem.items.length;
        });
        commit("CLEAR_PIZZA_ITEM", { rootState: rootState.cart, updateCart });
      } catch (err) {
        commit("CLEAR_PIZZA_ITEM_ERROR", err);
      }
    },
    plusCartItem({ commit, rootState, getters }, payLoad) {
      try {
        const addItem = produce(rootState.cart.pizzaItems, draft => {
          const cartItem = draft[payLoad.ID];
          const lastGroupItem = cartItem.items.length - 1;
          const addObj = [...cartItem.items, cartItem.items[lastGroupItem]];

          cartItem.totalPrice = getCartMethod.getTotalPrice(addObj, "price");
          cartItem.items = addObj;
        });
        const cartStatsUpdate = getters.getUpdateStatsCart(addItem);
        commit("PLUS_CART_ITEM", {
          rootState: rootState.cart,
          addItem,
          cartStatsUpdate
        });
      } catch (err) {
        commit("PLUS_PIZZA_CART_ITEM_ERROR", err);
      }
    },
    minusCartItem({ commit, rootState, getters }, payLoad) {
      try {
        const removeItem = produce(rootState.cart.pizzaItems, draft => {
          const cartItem = draft[payLoad.ID];
          if (cartItem.items.length > 1) {
            cartItem.items.pop();
            cartItem.totalPrice = getCartMethod.getTotalPrice(
              cartItem.items,
              "price"
            );
          } else {
            [...cartItem.items];
          }
        });
        const cartStatsUpdate = getters.getUpdateStatsCart(removeItem);
        commit("MINUS_CART_ITEM", {
          rootState: rootState.cart,
          removeItem,
          cartStatsUpdate
        });
      } catch (err) {
        commit("MINUS_CART_ITEM_ERROR", err);
      }
    }
  }
};
