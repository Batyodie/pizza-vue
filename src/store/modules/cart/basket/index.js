import getCartData from "@/func/getCartData";
import actions from "./action";
import mutations from "./mutation";
const getCartMethod = getCartData();
export default {
  state: {
    itemsError: null,
    itemError: null,
    plusItemError: null,
    minusItemError: null
  },
  getters: {
    getUpdateStatsCart: () => {
      return stateObj => {
        const pizzaObjData = getCartMethod.getTotalCout(stateObj);
        return {
          pizzaItemsCount: pizzaObjData.length,
          totalPrice: getCartMethod.getTotalPrice(pizzaObjData, "price")
        };
      };
    }
  },
  modules: {
    mutations,
    actions
  }
};
