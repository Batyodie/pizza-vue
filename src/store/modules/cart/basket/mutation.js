import {
  CLEAR_PIZZA_ITEMS,
  CLEAR_PIZZA_ITEMS_ERROR,
  CLEAR_PIZZA_ITEM,
  CLEAR_PIZZA_ITEM_ERROR,
  PLUS_CART_ITEM,
  PLUS_PIZZA_CART_ITEM_ERROR,
  MINUS_CART_ITEM,
  MINUS_PIZZA_CART_ITEM_ERROR
} from "../../../mutation-types";
export default {
  mutations: {
    [CLEAR_PIZZA_ITEMS](state, rootState) {
      state;
      const _s = rootState;
      _s.pizzaItems = {};
      _s.pizzaItemsCount = 0;
      _s.totalPrice = 0;
      _s.BasketItems = false;
    },
    [CLEAR_PIZZA_ITEMS_ERROR](state, error) {
      state.itemsError = error;
    },
    [CLEAR_PIZZA_ITEM](state, payLoad) {
      state;
      const _s = payLoad.rootState;
      const _update = payLoad.updateCart;

      _s.pizzaItems = _update.pizzaItems;
      _s.pizzaItemsCount -= _update.pizzaItemsCount;
      _s.totalPrice -= _update.totalPrice;
    },
    [CLEAR_PIZZA_ITEM_ERROR](state, error) {
      state.itemError = error;
    },
    [PLUS_CART_ITEM](state, payLoad) {
      state;
      const _s = payLoad.rootState;
      const _update = payLoad.addItem;
      const _updateStats = payLoad.cartStatsUpdate;

      _s.pizzaItems = _update;
      _s.pizzaItemsCount = _updateStats.pizzaItemsCount;
      _s.totalPrice = _updateStats.totalPrice;
    },
    [PLUS_PIZZA_CART_ITEM_ERROR](state, error) {
      state.plusItemError = error;
    },

    [MINUS_CART_ITEM](state, payLoad) {
      state;
      const _s = payLoad.rootState;
      const _update = payLoad.removeItem;
      const _updateStats = payLoad.cartStatsUpdate;

      _s.pizzaItems = _update;
      _s.pizzaItemsCount = _updateStats.pizzaItemsCount;
      _s.totalPrice = _updateStats.totalPrice;
    },
    [MINUS_PIZZA_CART_ITEM_ERROR](state, error) {
      state.plusItemError = error;
    }
  }
};
