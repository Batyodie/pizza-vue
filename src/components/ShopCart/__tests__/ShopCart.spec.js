import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";
import VModal from "vue-js-modal";
import basket from "@/store/modules/cart/basket";

jest.mock("@/store/modules/cart/basket");

import { ShopCart } from "@/components";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(
  VModal,
  { dialog: true },
  { componentName: "v-dialog-pay", dialog: true }
);

describe("Unit tests for ShopCart component", () => {
  // fake props
  const shopCart = [
    {
      id: 7,
      name: "Маргарита",
      imageUrl:
        "https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg",
      price: 450,
      size: 26,
      type: 0
    }
  ];

  const groupCartItems = [
    {
      id: 7,
      name: "Маргарита",
      imageUrl:
        "https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg",
      price: 450,
      size: 26,
      type: 0
    }
  ];
  const groupCartItemsPrice = groupCartItems.reduce(
    (sum, obj) => obj.price + sum,
    0
  );
  const index = 0;

  let store;
  let actions;
  // let mutations;

  beforeEach(() => {
    actions = {
      plusCartItem: jest.fn(),
      minusCartItem: jest.fn(),
      removePizzaItem: jest.fn(),
      removePizzaItems: jest.fn()
    };
    store = new Vuex.Store({
      state: basket.state,
      getters: basket.getters,
      actions
      // mutations,
    });
  });

  // create wrapper factory
  const wrapperFactory = (mouning = shallowMount, { propsData } = {}) => {
    return mouning(ShopCart, {
      localVue,
      store,
      propsData: {
        shopCart,
        groupCartItems,
        groupCartItemsPrice,
        index,
        ...propsData
      }
    });
  };
  it("should initialize correctly shopcart component", () => {
    const wrapper = wrapperFactory();
    // correctly Vue instance
    expect(wrapper.exists()).toBeTruthy();
    // We expect the wrapper to be the shopcart component
    expect(wrapper.is(ShopCart)).toBeTruthy();
    //  We expect the wrapper props to equal the fake props
    expect(wrapper.vm.shopCart).toEqual(shopCart);
    expect(wrapper.vm.groupCartItems).toEqual(groupCartItems);
    expect(wrapper.vm.groupCartItemsPrice).toEqual(groupCartItemsPrice);
    expect(wrapper.vm.index).toEqual(index);
  });

  it("should correctly render shopcart component", () => {
    const wrapper = wrapperFactory(mount);
    const renderSubTitle = `тонкое тесто, ${shopCart[0].size} см.`;
    const renerTotalPrice = `${groupCartItemsPrice} ₽`;

    // expect the shopcart img pizza to have correctly render image url
    expect(wrapper.find("picture > img").attributes("src")).toEqual(
      shopCart[0].imageUrl
    );
    // expect the shopcart to have correctly render title text
    expect(wrapper.find(".Title").text()).toEqual(shopCart[0].name);
    // expect the shopcart to have correctly render sub-title text
    expect(wrapper.find(".SubTitle").text()).toEqual(renderSubTitle);
    // expect the shopcart to have correctly render count quantity to group pizzas
    expect(wrapper.find(".Counter > span").text()).toEqual(
      String(groupCartItems.length)
    );
    // expect the shopcart to have correctly render total price count
    expect(wrapper.find(".TotalPrice").text()).toEqual(renerTotalPrice);
  });

  it("should match the snapshot ShopCart component", () => {
    const wrapper = wrapperFactory();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should correctly call shopcart actions plusCart to use button in comp", async () => {
    const wrapper = wrapperFactory();
    // find plus button to add group pizza item
    const counterBtn = wrapper.findAll(".CounterBtn");
    // click plus button
    await counterBtn.at(1).trigger("click");
    // expect to actions plus cart items to be called
    expect(actions.plusCartItem).toHaveBeenCalled();
  });
  it("should coorectly call shopcart actions munusCart to use button in comp", async () => {
    const wrapper = wrapperFactory();
    // find minus button to add group pizza item
    const counterBtn = wrapper.findAll(".CounterBtn");
    // click plus button
    await counterBtn.at(0).trigger("click");
    // expect to actions minus cart items to be called
    expect(actions.minusCartItem).toHaveBeenCalled();
  });

  it("should coorectly call shopcart actions removePizzaItem to use button in comp", async () => {
    const wrapper = wrapperFactory(mount);
    // find reset group items button
    const counterBtn = wrapper.find(".ResetBtn");
    // click reset btn
    await counterBtn.trigger("click");
    // find modal window button to accept deleteted group pizza items
    const addToResetBtn = wrapper.findAll(".vue-dialog-button");
    // click to accept delete button
    await addToResetBtn.at(1).trigger("click");
    // expect to actions removePizzaItem to be called
    expect(actions.removePizzaItem).toHaveBeenCalled();
  });

  it("should coorectly call shopcart actions removePizzaItem to use button in comp", async () => {
    const wrapper = wrapperFactory(mount);
    // find reset group items button
    const counterBtn = wrapper.find(".ResetBtn");
    // click reset btn
    await counterBtn.trigger("click");
    // find modal window button to close deleteted group pizza items
    const addToResetBtn = wrapper.findAll(".vue-dialog-button");
    // click to close delete button
    await addToResetBtn.at(0).trigger("click");
    // expect to actions removePizzaItem to not have be called
    expect(actions.removePizzaItem).not.toHaveBeenCalled();
  });
});
