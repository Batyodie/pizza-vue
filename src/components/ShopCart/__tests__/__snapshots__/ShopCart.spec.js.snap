// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for ShopCart component should match the snapshot ShopCart component 1`] = `
<div class="ShopCart">
  <picture>
    <source srcset="https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg" width="80" height="80" type="image/svg"> <img src="https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg" width="80" height="80" alt="ShopCart pizza image">
  </picture>
  <div class="Typh">
    <h3 class="Title">Маргарита</h3>
    <p class="SubTitle">
      тонкое тесто, 26 см.
    </p>
  </div>
  <div class="Counter">
    <button-stub class="CounterBtn"><template></template></button-stub> <span style="margin: 0px 12px 0px 12px;">
      1
    </span>
    <button-stub class="CounterBtn"><template></template></button-stub>
  </div>
  <div class="TotalContainer">
    <div class="TotalPrice"><span> 450 ₽ </span></div>
  </div>
  <div class="Reset">
    <button-stub class="ResetBtn"><template></template></button-stub>
    <v-dialog-stub width="400" clicktoclose="true"></v-dialog-stub>
  </div>
</div>
`;
